### PR TITLE
Port SDL2 implementation and interop methods from MonoGame

### PR DESCRIPTION
**Pull request by sr229**
This is a scaffold for the upcoming SDL2 backend, along with the interop handles used by MonoGame that we might need for our use case.

### CODE REVIEW

**Review by LeNitrous — DISMISSED** [review]
I don't feel confident with PR at all with all the bloat it has. Here are some examples:
- Why do we need to reimplement native library loading when there's `NativeLibrary` in .NET 6
- Why do we need to reimplement high resolution timing when `Stopwatch` supports it
- Use fixed size char buffers when dealing with strings to avoid allocations
- Generally has too much bloat

File: source/extensions/Sekai.Interop.Common/FuncLoader.Desktop.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+// Licensed under MIT. See LICENSE for details.
+
+// Contains code from MonoGame.
+// Copyright 2021 the MonoGame Team.
+// Licensed under the Microsoft Public License and MIT.
+// See https://github.com/MonoGame/MonoGame/blob/develop/LICENSE.txt.
+using System.Runtime.InteropServices;
+
+namespace Sekai.Interop.Common;
+public class FuncLoader
```
Comment (adryzz) [inline source/extensions/Sekai.Interop.Common/FuncLoader.Desktop.cs:11]: [System.Runtime.InteropServices.NativeLibrary](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary?view=net-6.0) seems to do what is written here

File: source/extensions/Sekai.Interop.Common/FuncLoader.Desktop.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        [DllImport("/usr/lib/libSystem.dylib")]
+        public static extern IntPtr dlsym(IntPtr handle, string symbol);
+    }
+
+    private const int rtld_lazy = 0x0001;
+
+    public static IntPtr LoadLibraryExt(string libname)
+    {
+        var ret = IntPtr.Zero;
+        string? assemblyLocation = Path.GetDirectoryName(typeof(FuncLoader).Assembly.Location) ?? "./";
```
Comment (adryzz) [inline source/extensions/Sekai.Interop.Common/FuncLoader.Desktop.cs:45]:
```suggestion
        string? assemblyLocation = Path.GetDirectoryName(typeof(FuncLoader).Assembly.Location) ?? Environment.CurrentDirectory;
```

File: source/extensions/Sekai.Interop.Common/ReflectionHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        public static int Get()
+        {
+            return sizeOf;
+        }
+    }
+
+    /// <summary>
+    /// Fallback handler for Marshal.SizeOf(type)
+    /// </summary>
+    internal static int ManagedSizeOf(Type type)
```
Comment (adryzz) [inline source/extensions/Sekai.Interop.Common/ReflectionHelpers.cs:35]: in which cases you'll have a structure of an unknown size to pass to a native function?

File: source/extensions/Sekai.Interop.Common/TimerHelper.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+// Licensed under MIT. See LICENSE for details.
+
+// Contains code from MonoGame.
+// Copyright 2021 the MonoGame Team.
+// Licensed under the Microsoft Public License and MIT.
+// See https://github.com/MonoGame/MonoGame/blob/develop/LICENSE.txt.
+using System.Runtime.InteropServices;
+
+namespace Sekai.Interop.Common;
+internal static class TimerHelper
```
Comment (adryzz) [inline source/extensions/Sekai.Interop.Common/TimerHelper.cs:11]: [System.Timers.Timer](https://docs.microsoft.com/en-us/dotnet/api/system.timers.timer?view=net-6.0) is already high resolution, and if you need more look at OTD's native timer implementation. [here](https://github.com/OpenTabletDriver/OpenTabletDriver/tree/master/OpenTabletDriver.Desktop/Interop/Timer) and [here for macos](https://github.com/OpenTabletDriver/OpenTabletDriver/blob/80b98ad170a809399b89aba48387f95d745f3b01/OpenTabletDriver.Desktop/Interop/Timer/MacOSTimer.cs)

**Review by adryzz — CHANGES_REQUESTED** [review]

File: source/extensions/Sekai.Interop.Common/FuncLoader.Desktop.cs (hunk 1 above)
Comment (sr229) [reply source/extensions/Sekai.Interop.Common/FuncLoader.Desktop.cs:11]: Lemme see if I can reimplement some of the methods using NativeLibrary. At best when I checked this, this seems to do some stuff better than `NativeLibrary` implementation wise.

File: source/extensions/Sekai.Interop.Common/TimerHelper.cs (hunk 4 above)
Comment (sr229) [reply source/extensions/Sekai.Interop.Common/TimerHelper.cs:11]: Removed on my local copy, will push that.

File: source/extensions/Sekai.Interop.Common/ReflectionHelpers.cs (hunk 3 above)
Comment (sr229) [reply source/extensions/Sekai.Interop.Common/ReflectionHelpers.cs:35]: I'm assuming they were accounting for cases where SizeOf might be undefined, though I believe this was done for Mono, which we can remove.

**Comment by sr229** [comment]
Trimmed down most of the things that didn't need to be there, but I have to keep some methods since `NativeLibrary` isn't a full "drop-in" replacement for it.

**Review by adryzz — APPROVED** [review]
should be fine

**Review by LeNitrous — COMMENTED** [review]
I suppose the intent here is to get the latest rev of SDL as soon as possible. If that were the case, we'd end up having to build our own binaries.

**Comment by sr229** [comment]
> I suppose the intent here is to get the latest rev of SDL as soon as possible. If that were the case, we'd end up having to build our own binaries.

That was the main intent of why we're doing our own managed wrapper based from MonoGame's implementation so we can ship the latest SDL. Of course we will still need to do QoLF but that's for future revisions.

> * We don't need `FuncLoader`. .NET's `NativeLibrary.Load` has an overload that handles search path lookups for you.
> * `Sekai.Interop.Desktop` isn't needed when SDL is the only library using it. We can revisit this at a later time.

Agreed, I kept the most-used methods and just shimmed them with `NativeLibrary.TryLoad()` so they behave the same way, but in the end of the day we will need to convert everything to `NativeLibrary`.

**Review by LeNitrous — APPROVED** [review]

**Pull request merged**